docs/proposal/20200506-single-controller-multitenancy.md

```diff
@@ -0,0 +1,903 @@
+
+Currently, without a custom deployment, role assumption in CAPA can only take place using a metadata proxy such as
+[kiam][kiam], or using OIDC based service account token volume projection, requiring Kubernetes 1.16, and special setup
+of certificates with an appropriate OIDC trust.
+
+It is technically possible to provide an IAM role via a shared configuration file, but the UX has been found to be
+complicated, and additional breaking code changes have been suggested to resolve these issues.
+
+In addition, where role assumption is configured, that role is used for the entire lifetime of the CAPA deployment. This
+also means that an AWSCluster could be broken if the instance of CAPA that created it is misconfigured for another set
+of credentials.
+
+This proposal outlines new capabilities for CAPA to use IAM role assumption to assume a permission set in a different
+AWS account, at runtime, on a per-cluster basis. The proposed changes would be fully backwards compatible and maintain
+the existing behavior with no changes to user configuration required.
+
+
+## Motivation
+
+For large organizations, especially highly-regulated organizations, there is a need to be able to perform separate
+duties at various levels of infrastructure - permissions, networks and accounts. VPC sharing is a model which provides
+separation at the AWS account level. Within this model it is appropriate for tooling running within the 'management' account
+to manage infrastructure within the 'workload' accounts, which requires a principal in the management account which can
+assume a principal within a workload account. For CAPA to be most useful within these organizations it will need to
+support multi-account models.
+
+Some organizations may also delegate the management of clusters to another third-party. In that case, the boundary
+between organizations needs to be secured, and issues such as confused deputy scenarios need to be mitigated.
+
+Because a single deployment of the CAPA operator may reconcile many different clusters in its lifetime, it is necessary
+to modify the CAPA operator to scope its AWS client instances to within the reconciliation process.
+
+It follows that an organization may wish to provision control planes and worker groups in separate accounts (including
+each worker group in a separate account). There is a desire to support this configuration also.
+
+AWS provides a number of mechanisms for assume roles across account boundaries:
+
+* Role assumption using credentials provided in environment variables
+* Role assumption using credentials provided by a metadata service, e.g. IMDS or ECS
+* Role assumption using an OIDC JWT token
+* Role assumption using explicitly provided credentials
+
+The first three methods are supported per instance of the CAPA controller today. Of special note is OIDC, which was
+introduced in 2019 as a way to leverage projected service account tokens from Kubernetes to assume role. This allows for
```

Comment:
I think if we can add a field to recognize role ARN in CRD. In this way, we can switch to different account role using the same JWT token.